[GSoC] Added Uploaded status table in room database.

4D17Y4 opened the pull request:
Added Uploaded Status table.
Updated the room database file.

Tasks :

- [x] Query original Image.
- [x] Documentation.

nicolas-raoul commented on app/src/main/java/fr/free/nrw/commons/customselector/database/UploadedDao.kt:
```diff
@@ -0,0 +1,112 @@
+     * Get All entries from the uploaded status table.
+     */
+    @Query("SELECT * FROM uploaded_table")
+    abstract suspend fun getAll() : List<UploadedStatus>
```
Do we really need this one?

4D17Y4 replied:
Just for debugging, I'll clear it off as soon as I finish working on this table.

nicolas-raoul commented:
Looking good and running fine so far :-)

4D17Y4 commented:
Ready for review 
@nicolas-raoul @madhurgupta10

4D17Y4 commented:
Should I remove all those unused Dao methods?

nicolas-raoul commented:
I would suggest commenting out methods that you think will not be needed. They can be removed during the review before merging to master.

Using the newest code I just got a crash when scrolling down a lot (I have thousands of pictures on my phone even though I move them to my PC every few months):
```
06-25 12:08:11.245 24296 24629 I FileUtils: File SHA1: b7eea74b74661c9e52998bc8667453884a61e57f
06-25 12:08:11.329 24296 24581 I FileUtils: File SHA1: 9647931e7ab82cd183dee92d09e863494ea5c3e0
06-25 12:08:11.345 24296 24296 E ACRA    : ACRA caught a KotlinNullPointerException for fr.free.nrw.commons
06-25 12:08:11.345 24296 24296 E ACRA    : kotlin.KotlinNullPointerException
06-25 12:08:11.345 24296 24296 E ACRA    : 	at fr.free.nrw.commons.customselector.ui.selector.ImageLoader$queryAndSetView$1.invokeSuspend(ImageLoader.kt:119)
06-25 12:08:11.345 24296 24296 E ACRA    : 	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
06-25 12:08:11.345 24296 24296 E ACRA    : 	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
06-25 12:08:11.345 24296 24296 E ACRA    : 	at android.os.Handler.handleCallback(Handler.java:873)
06-25 12:08:11.345 24296 24296 E ACRA    : 	at android.os.Handler.dispatchMessage(Handler.java:99)
06-25 12:08:11.345 24296 24296 E ACRA    : 	at android.os.Looper.loop(Looper.java:214)
06-25 12:08:11.345 24296 24296 E ACRA    : 	at android.app.ActivityThread.main(ActivityThread.java:7032)
06-25 12:08:11.345 24296 24296 E ACRA    : 	at java.lang.reflect.Method.invoke(Native Method)
06-25 12:08:11.345 24296 24296 E ACRA    : 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:494)
06-25 12:08:11.345 24296 24296 E ACRA    : 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:965)
06-25 12:08:11.351 24296 24581 I FileUtils: File SHA1: 73fb5f9614f591a59f49d95be57df1b64baf202d
06-25 12:08:11.375 24296 24631 I FileUtils: File SHA1: 5c7f170377d2c3d20f175c3940422222e34f8736
```

4D17Y4 commented:
Handled the exception.
I changed the default value of the result to null and did not check before using it. 
Should work as expected now.

madhurgupta10 commented on app/src/main/java/fr/free/nrw/commons/customselector/ui/selector/ImageLoader.kt:
```diff
@@ -62,22 +69,54 @@ class ImageLoader @Inject constructor(
+                }
             }
-            if(mapHolderImage[holder] == image) {
+            if(mapHolderImage[holder] == image && result!=null) {
```
@4D17Y4 In Kotlin there is a much better way to handle null using `let`

madhurgupta10 commented on app/src/main/java/fr/free/nrw/commons/customselector/database/UploadedDao.kt:
```diff
@@ -0,0 +1,112 @@
+        runBlocking {
+            launch(Dispatchers.IO) {
+                uploadedStatus.lastUpdated = Calendar.getInstance().time as Date?
+                insert(uploadedStatus)
```
suspend functions can be handled using `async await` easily

madhurgupta10 commented on app/src/main/java/fr/free/nrw/commons/customselector/ui/selector/ImageLoader.kt:
```diff
@@ -62,22 +69,54 @@ class ImageLoader @Inject constructor(
+                        // already uploaded image.
+                        result = true
+                    }
+                    else if(uploadedStatus.lastUpdated!!.date >=
```
This can invoke a NPE here, better to handle it gracefully

4D17Y4 replied:
Done.

nicolas-raoul commented:
In airplane mode I get this crash:
```
06-25 12:38:04.378  5621  5803 I clp-JNI : Add pinfo.  (channel) f67a41b fr.free.nrw.commons/fr.free.nrw.commons.customselector.ui.selector.CustomSelectorActivity (server) (pkg) AppWindowToken{3fa2ccc token=Token{cb696ff ActivityRecord{d644e1e u0 fr.free.nrw.commons/.customselector.ui.selector.CustomSelectorActivity t25284}}} 31190 10401 (action) 0
06-25 12:38:04.383 31190 31190 D ViewRootImpl@c053720[CustomSelectorActivity]: ViewPostIme pointer 0
06-25 12:38:04.428  5621  5803 I clp-JNI : Add pinfo.  (channel) f67a41b fr.free.nrw.commons/fr.free.nrw.commons.customselector.ui.selector.CustomSelectorActivity (server) (pkg) AppWindowToken{3fa2ccc token=Token{cb696ff ActivityRecord{d644e1e u0 fr.free.nrw.commons/.customselector.ui.selector.CustomSelectorActivity t25284}}} 31190 10401 (action) 1
06-25 12:38:04.430 31190 31190 D ViewRootImpl@c053720[CustomSelectorActivity]: ViewPostIme pointer 1
06-25 12:38:04.609 31190 31495 I FileUtils: File SHA1: fe992e70e82c1a9d7750f8477f96060f150016cb
06-25 12:38:04.616 31190 31497 I FileUtils: File SHA1: 85d48ba2c18a17f89e9c0489d78d36babfdd32a7
06-25 12:38:04.620 31190 31496 I FileUtils: File SHA1: 119716430dac844839bf58a10f87868771b81997
06-25 12:38:04.620 31190 31502 I FileUtils: File SHA1: 2307262d77324b78db59e5e01b4a4b6803706525
06-25 12:38:04.624 31190 31505 I FileUtils: File SHA1: a5a30b2bd5a9eacd3f27b9f970d5476ec54ca509
06-25 12:38:04.624 31190 31508 I FileUtils: File SHA1: 16c294072ed5595fcb5ea58056063b3852006670
06-25 12:38:04.625 31190 31507 I FileUtils: File SHA1: 8eaaed87cbf2cce0dbb71f2079e7df9e942f5420
06-25 12:38:04.640 31190 31503 I FileUtils: File SHA1: 16890a6a5c4b7d8e58694289567dabb3e5b7833f
06-25 12:38:04.662 31190 31495 D FileMetadataUtils: Retuning tags for pref:Author
06-25 12:38:04.672 31190 31497 D FileMetadataUtils: Retuning tags for pref:Author
06-25 12:38:04.674 31190 31508 D FileMetadataUtils: Retuning tags for pref:Author
06-25 12:38:04.674 31190 31508 D FileProcessor: Checking for tag: Artist
06-25 12:38:04.674 31190 31508 D FileProcessor: Checking for tag: CameraOwnerName
06-25 12:38:04.675 31190 31508 D FileMetadataUtils: Retuning tags for pref:Copyright
06-25 12:38:04.675 31190 31508 D FileProcessor: Checking for tag: Copyright
06-25 12:38:04.675 31190 31508 D FileMetadataUtils: Retuning tags for pref:Camera Model
06-25 12:38:04.676 31190 31508 D FileProcessor: Checking for tag: Make
06-25 12:38:04.676 31190 31508 D FileProcessor: Checking for tag: Model
06-25 12:38:04.676 31190 31508 D FileMetadataUtils: Retuning tags for pref:Lens Model
06-25 12:38:04.677 31190 31508 D FileProcessor: Checking for tag: LensMake
06-25 12:38:04.677 31190 31508 D FileProcessor: Checking for tag: LensModel
06-25 12:38:04.677 31190 31508 D FileProcessor: Checking for tag: LensSpecification
06-25 12:38:04.677 31190 31508 D FileMetadataUtils: Retuning tags for pref:Serial Numbers
06-25 12:38:04.678 31190 31495 D FileProcessor: Checking for tag: Artist
06-25 12:38:04.678 31190 31508 D FileProcessor: Checking for tag: BodySerialNumber
06-25 12:38:04.678 31190 31495 D FileProcessor: Checking for tag: CameraOwnerName
06-25 12:38:04.679 31190 31508 D FileProcessor: Checking for tag: LensSerialNumber
06-25 12:38:04.679 31190 31495 D FileMetadataUtils: Retuning tags for pref:Copyright
06-25 12:38:04.679 31190 31508 D FileMetadataUtils: Retuning tags for pref:Software
06-25 12:38:04.679 31190 31495 D FileProcessor: Checking for tag: Copyright
06-25 12:38:04.680 31190 31508 D FileProcessor: Checking for tag: Software
06-25 12:38:04.680 31190 31495 D FileMetadataUtils: Retuning tags for pref:Camera Model
06-25 12:38:04.680 31190 31495 D FileProcessor: Checking for tag: Make
06-25 12:38:04.680 31190 31495 D FileProcessor: Checking for tag: Model
06-25 12:38:04.681 31190 31495 D FileMetadataUtils: Retuning tags for pref:Lens Model
06-25 12:38:04.681 31190 31495 D FileProcessor: Checking for tag: LensMake
06-25 12:38:04.681 31190 31495 D FileProcessor: Checking for tag: LensModel
06-25 12:38:04.682 31190 31495 D FileProcessor: Checking for tag: LensSpecification
06-25 12:38:04.684 31190 31495 D FileMetadataUtils: Retuning tags for pref:Serial Numbers
06-25 12:38:04.684 31190 31497 D FileProcessor: Checking for tag: Artist
06-25 12:38:04.684 31190 31497 D FileProcessor: Checking for tag: CameraOwnerName
06-25 12:38:04.685 31190 31496 D FileMetadataUtils: Retuning tags for pref:Author
06-25 12:38:04.685 31190 31507 D FileMetadataUtils: Retuning tags for pref:Author
06-25 12:38:04.687 31190 31496 D FileProcessor: Checking for tag: Artist
06-25 12:38:04.687 31190 31497 D FileMetadataUtils: Retuning tags for pref:Copyright
06-25 12:38:04.687 31190 31495 D FileProcessor: Checking for tag: BodySerialNumber
06-25 12:38:04.688 31190 31507 D FileProcessor: Checking for tag: Artist
06-25 12:38:04.688 31190 31507 D FileProcessor: Checking for tag: CameraOwnerName
06-25 12:38:04.689 31190 31507 D FileMetadataUtils: Retuning tags for pref:Copyright
06-25 12:38:04.689 31190 31495 D FileProcessor: Checking for tag: LensSerialNumber
06-25 12:38:04.689 31190 31507 D FileProcessor: Checking for tag: Copyright
06-25 12:38:04.690 31190 31507 D FileMetadataUtils: Retuning tags for pref:Camera Model
06-25 12:38:04.690 31190 31507 D FileProcessor: Checking for tag: Make
06-25 12:38:04.690 31190 31495 D FileMetadataUtils: Retuning tags for pref:Software
06-25 12:38:04.690 31190 31507 D FileProcessor: Checking for tag: Model
06-25 12:38:04.691 31190 31507 D FileMetadataUtils: Retuning tags for pref:Lens Model
06-25 12:38:04.691 31190 31507 D FileProcessor: Checking for tag: LensMake
06-25 12:38:04.691 31190 31495 D FileProcessor: Checking for tag: Software
06-25 12:38:04.691 31190 31507 D FileProcessor: Checking for tag: LensModel
06-25 12:38:04.692 31190 31507 D FileProcessor: Checking for tag: LensSpecification
06-25 12:38:04.692 31190 31507 D FileMetadataUtils: Retuning tags for pref:Serial Numbers
06-25 12:38:04.693 31190 31507 D FileProcessor: Checking for tag: BodySerialNumber
06-25 12:38:04.694 31190 31496 D FileProcessor: Checking for tag: CameraOwnerName
06-25 12:38:04.694 31190 31497 D FileProcessor: Checking for tag: Copyright
06-25 12:38:04.694 31190 31496 D FileMetadataUtils: Retuning tags for pref:Copyright
06-25 12:38:04.694 31190 31497 D FileMetadataUtils: Retuning tags for pref:Camera Model
06-25 12:38:04.694 31190 31496 D FileProcessor: Checking for tag: Copyright
06-25 12:38:04.695 31190 31497 D FileProcessor: Checking for tag: Make
06-25 12:38:04.695 31190 31496 D FileMetadataUtils: Retuning tags for pref:Camera Model
06-25 12:38:04.695 31190 31497 D FileProcessor: Checking for tag: Model
06-25 12:38:04.695 31190 31496 D FileProcessor: Checking for tag: Make
06-25 12:38:04.695 31190 31497 D FileMetadataUtils: Retuning tags for pref:Lens Model
06-25 12:38:04.695 31190 31508 I FileUtils: File SHA1: 2123249a2991f5e0254c3fa4fd16d07681db5d90
06-25 12:38:04.696 31190 31496 D FileProcessor: Checking for tag: Model
06-25 12:38:04.696 31190 31497 D FileProcessor: Checking for tag: LensMake
06-25 12:38:04.696 31190 31496 D FileMetadataUtils: Retuning tags for pref:Lens Model
06-25 12:38:04.696 31190 31507 D FileProcessor: Checking for tag: LensSerialNumber
06-25 12:38:04.696 31190 31507 D FileMetadataUtils: Retuning tags for pref:Software
06-25 12:38:04.696 31190 31496 D FileProcessor: Checking for tag: LensMake
06-25 12:38:04.697 31190 31507 D FileProcessor: Checking for tag: Software
06-25 12:38:04.697 31190 31496 D FileProcessor: Checking for tag: LensModel
06-25 12:38:04.697 31190 31496 D FileProcessor: Checking for tag: LensSpecification
06-25 12:38:04.697 31190 31496 D FileMetadataUtils: Retuning tags for pref:Serial Numbers
06-25 12:38:04.698 31190 31496 D FileProcessor: Checking for tag: BodySerialNumber
06-25 12:38:04.699 31190 31497 D FileProcessor: Checking for tag: LensModel
06-25 12:38:04.700 31190 31497 D FileProcessor: Checking for tag: LensSpecification
06-25 12:38:04.700 31190 31496 D FileProcessor: Checking for tag: LensSerialNumber
06-25 12:38:04.703 31190 31508 D OkHttp  : --> GET https://commons.wikimedia.org/w/api.php?action=query&format=json&formatversion=2&list=allimages&aisha1=2123249a2991f5e0254c3fa4fd16d07681db5d90
06-25 12:38:04.704 31190 31497 D FileMetadataUtils: Retuning tags for pref:Serial Numbers
06-25 12:38:04.704 31190 31505 D FileMetadataUtils: Retuning tags for pref:Author
06-25 12:38:04.705 31190 31505 D FileProcessor: Checking for tag: Artist
06-25 12:38:04.705 31190 31505 D FileProcessor: Checking for tag: CameraOwnerName
06-25 12:38:04.705 31190 31505 D FileMetadataUtils: Retuning tags for pref:Copyright
06-25 12:38:04.706 31190 31497 D FileProcessor: Checking for tag: BodySerialNumber
06-25 12:38:04.708 31190 31505 D FileProcessor: Checking for tag: Copyright
06-25 12:38:04.708 31190 31505 D FileMetadataUtils: Retuning tags for pref:Camera Model
06-25 12:38:04.708 31190 31505 D FileProcessor: Checking for tag: Make
06-25 12:38:04.709 31190 31496 D FileMetadataUtils: Retuning tags for pref:Software
06-25 12:38:04.712 31190 31497 D FileProcessor: Checking for tag: LensSerialNumber
06-25 12:38:04.712 31190 31505 D FileProcessor: Checking for tag: Model
06-25 12:38:04.712 31190 31508 D CookieManager: Domain:commons.wikimedia.org
06-25 12:38:04.713 31190 31496 D FileProcessor: Checking for tag: Software
06-25 12:38:04.714 31190 31505 D FileMetadataUtils: Retuning tags for pref:Lens Model
06-25 12:38:04.714 31190 31505 D FileProcessor: Checking for tag: LensMake
06-25 12:38:04.714 31190 31507 I FileUtils: File SHA1: 958ec3dfbec43f33fc9573cb6dc18faed9148e81
06-25 12:38:04.714 31190 31497 D FileMetadataUtils: Retuning tags for pref:Software
06-25 12:38:04.716 31190 31505 D FileProcessor: Checking for tag: LensModel
06-25 12:38:04.717 31190 31507 D OkHttp  : --> GET https://commons.wikimedia.org/w/api.php?action=query&format=json&formatversion=2&list=allimages&aisha1=958ec3dfbec43f33fc9573cb6dc18faed9148e81
06-25 12:38:04.717 31190 31503 D FileMetadataUtils: Retuning tags for pref:Author
06-25 12:38:04.718 31190 31503 D FileProcessor: Checking for tag: Artist
06-25 12:38:04.719 31190 31508 D OkHttp  : <-- HTTP FAILED: java.net.UnknownHostException: Unable to resolve host "commons.wikimedia.org": No address associated with hostname
06-25 12:38:04.719 31190 31495 I FileUtils: File SHA1: eceabf599f6d8fbe328e2e82c4ee89af07919888
06-25 12:38:04.720 31190 31497 D FileProcessor: Checking for tag: Software
06-25 12:38:04.720 31190 31503 D FileProcessor: Checking for tag: CameraOwnerName
06-25 12:38:04.722 31190 31495 D OkHttp  : --> GET https://commons.wikimedia.org/w/api.php?action=query&format=json&formatversion=2&list=allimages&aisha1=eceabf599f6d8fbe328e2e82c4ee89af07919888
06-25 12:38:04.724 31190 31507 D CookieManager: Domain:commons.wikimedia.org
06-25 12:38:04.725 31190 31495 D CookieManager: Domain:commons.wikimedia.org
06-25 12:38:04.727 31190 31507 D OkHttp  : <-- HTTP FAILED: java.net.UnknownHostException: Unable to resolve host "commons.wikimedia.org": No address associated with hostname
06-25 12:38:04.730 31190 31503 D FileMetadataUtils: Retuning tags for pref:Copyright
06-25 12:38:04.730 31190 31502 D FileMetadataUtils: Retuning tags for pref:Author
06-25 12:38:04.732 31190 31496 I FileUtils: File SHA1: 1027ab208690bfcdda01bdc99cb4624dd275e242
06-25 12:38:04.732 31190 31505 D FileProcessor: Checking for tag: LensSpecification
06-25 12:38:04.737 31190 31496 D OkHttp  : --> GET https://commons.wikimedia.org/w/api.php?action=query&format=json&formatversion=2&list=allimages&aisha1=1027ab208690bfcdda01bdc99cb4624dd275e242
06-25 12:38:04.742 31190 31495 D OkHttp  : <-- HTTP FAILED: java.net.UnknownHostException: Unable to resolve host "commons.wikimedia.org": No address associated with hostname
06-25 12:38:04.742 31190 31190 E ACRA    : ACRA caught a RuntimeException for fr.free.nrw.commons
06-25 12:38:04.743 31190 31496 D CookieManager: Domain:commons.wikimedia.org
06-25 12:38:04.746 31190 31507 I FileUtils: File SHA1: aff53229f97413e4816fb775f343205ecc367572
06-25 12:38:04.746 31190 31497 I FileUtils: File SHA1: 5bff0ee59f7b5b4f716b8182519ba3804eea9c4a
06-25 12:38:04.748 31190 31508 I FileUtils: File SHA1: 22bfc468a5d254c7eb664e317a356f9d78ca0743
06-25 12:38:04.749 31190 31505 D FileMetadataUtils: Retuning tags for pref:Serial Numbers
06-25 12:38:04.749 31190 31503 D FileProcessor: Checking for tag: Copyright
06-25 12:38:04.750 31190 31502 D FileProcessor: Checking for tag: Artist
06-25 12:38:04.752 31190 31505 D FileProcessor: Checking for tag: BodySerialNumber
06-25 12:38:04.753 31190 31503 D FileMetadataUtils: Retuning tags for pref:Camera Model
06-25 12:38:04.754 31190 31497 D OkHttp  : --> GET https://commons.wikimedia.org/w/api.php?action=query&format=json&formatversion=2&list=allimages&aisha1=5bff0ee59f7b5b4f716b8182519ba3804eea9c4a
06-25 12:38:04.754 31190 31496 D OkHttp  : <-- HTTP FAILED: java.net.UnknownHostException: Unable to resolve host "commons.wikimedia.org": No address associated with hostname
06-25 12:38:04.755 31190 31502 D FileProcessor: Checking for tag: CameraOwnerName
06-25 12:38:04.755 31190 31505 D FileProcessor: Checking for tag: LensSerialNumber
06-25 12:38:04.758 31190 31495 I FileUtils: File SHA1: 0b0830a53c042363dc89b3c78102f0b4178a6c08
06-25 12:38:04.759 31190 31497 D CookieManager: Domain:commons.wikimedia.org
06-25 12:38:04.762 31190 31505 D FileMetadataUtils: Retuning tags for pref:Software
06-25 12:38:04.762 31190 31503 D FileProcessor: Checking for tag: Make
06-25 12:38:04.762 31190 31502 D FileMetadataUtils: Retuning tags for pref:Copyright
06-25 12:38:04.766 31190 31503 D FileProcessor: Checking for tag: Model
06-25 12:38:04.768 31190 31502 D FileProcessor: Checking for tag: Copyright
06-25 12:38:04.769 31190 31505 D FileProcessor: Checking for tag: Software
06-25 12:38:04.773 31190 31502 D FileMetadataUtils: Retuning tags for pref:Camera Model
06-25 12:38:04.773 31190 31503 D FileMetadataUtils: Retuning tags for pref:Lens Model
06-25 12:38:04.782 31190 31496 I FileUtils: File SHA1: 97f683a3c49e4725dacf8ae13482b4553dd68e25
06-25 12:38:04.782 31190 31497 D OkHttp  : <-- HTTP FAILED: java.net.UnknownHostException: Unable to resolve host "commons.wikimedia.org": No address associated with hostname
06-25 12:38:04.785 31190 31502 D FileProcessor: Checking for tag: Make
06-25 12:38:04.790 31190 31503 D FileProcessor: Checking for tag: LensMake
06-25 12:38:04.797 31190 31503 D FileProcessor: Checking for tag: LensModel
06-25 12:38:04.799 31190 31502 D FileProcessor: Checking for tag: Model
06-25 12:38:04.802 31190 31507 D FileMetadataUtils: Retuning tags for pref:Author
06-25 12:38:04.805 31190 31503 D FileProcessor: Checking for tag: LensSpecification
06-25 12:38:04.805 31190 31502 D FileMetadataUtils: Retuning tags for pref:Lens Model
06-25 12:38:04.810 31190 31507 D FileProcessor: Checking for tag: Artist
06-25 12:38:04.814 31190 31503 D FileMetadataUtils: Retuning tags for pref:Serial Numbers
06-25 12:38:04.815 31190 31508 D FileMetadataUtils: Retuning tags for pref:Author
06-25 12:38:04.816 31190 31502 D FileProcessor: Checking for tag: LensMake
06-25 12:38:04.816 31190 31497 I FileUtils: File SHA1: 03e3912a38f3976eceb2ab70bea53e1fbf01f9d3
06-25 12:38:04.818 31190 31495 D FileMetadataUtils: Retuning tags for pref:Author
06-25 12:38:04.818 31190 31503 D FileProcessor: Checking for tag: BodySerialNumber
06-25 12:38:04.819 31190 31508 D FileProcessor: Checking for tag: Artist
06-25 12:38:04.819 31190 31507 D FileProcessor: Checking for tag: CameraOwnerName
06-25 12:38:04.821 31190 31502 D FileProcessor: Checking for tag: LensModel
06-25 12:38:04.822 31190 31503 D FileProcessor: Checking for tag: LensSerialNumber
06-25 12:38:04.823 31190 31507 D FileMetadataUtils: Retuning tags for pref:Copyright
06-25 12:38:04.824 31190 31508 D FileProcessor: Checking for tag: CameraOwnerName
06-25 12:38:04.825 31190 31503 D FileMetadataUtils: Retuning tags for pref:Software
06-25 12:38:04.825 31190 31502 D FileProcessor: Checking for tag: LensSpecification
06-25 12:38:04.825 31190 31495 D FileProcessor: Checking for tag: Artist
06-25 12:38:04.826 31190 31507 D FileProcessor: Checking for tag: Copyright
06-25 12:38:04.827 31190 31496 D FileMetadataUtils: Retuning tags for pref:Author
06-25 12:38:04.828 31190 31502 D FileMetadataUtils: Retuning tags for pref:Serial Numbers
06-25 12:38:04.829 31190 31508 D FileMetadataUtils: Retuning tags for pref:Copyright
06-25 12:38:04.829 31190 31507 D FileMetadataUtils: Retuning tags for pref:Camera Model
06-25 12:38:04.834 31190 31503 D FileProcessor: Checking for tag: Software
06-25 12:38:04.835 31190 31495 D FileProcessor: Checking for tag: CameraOwnerName
06-25 12:38:04.835 31190 31508 D FileProcessor: Checking for tag: Copyright
06-25 12:38:04.837 31190 31495 D FileMetadataUtils: Retuning tags for pref:Copyright
06-25 12:38:04.846 31190 31496 D FileProcessor: Checking for tag: Artist
06-25 12:38:04.846 31190 31505 I FileUtils: File SHA1: 390ca7f8d8a42f14ab8afdd8c2fe21bcca99b1b4
06-25 12:38:04.847 31190 31507 D FileProcessor: Checking for tag: Make
06-25 12:38:04.847 31190 31507 D FileProcessor: Checking for tag: Model
06-25 12:38:04.848 31190 31502 D FileProcessor: Checking for tag: BodySerialNumber
06-25 12:38:04.848 31190 31507 D FileMetadataUtils: Retuning tags for pref:Lens Model
06-25 12:38:04.848 31190 31507 D FileProcessor: Checking for tag: LensMake
06-25 12:38:04.849 31190 31507 D FileProcessor: Checking for tag: LensModel
06-25 12:38:04.849 31190 31507 D FileProcessor: Checking for tag: LensSpecification
06-25 12:38:04.849 31190 31507 D FileMetadataUtils: Retuning tags for pref:Serial Numbers
06-25 12:38:04.849 31190 31505 D OkHttp  : --> GET https://commons.wikimedia.org/w/api.php?action=query&format=json&formatversion=2&list=allimages&aisha1=390ca7f8d8a42f14ab8afdd8c2fe21bcca99b1b4
06-25 12:38:04.850 31190 31502 D FileProcessor: Checking for tag: LensSerialNumber
06-25 12:38:04.850 31190 31508 D FileMetadataUtils: Retuning tags for pref:Camera Model
06-25 12:38:04.853 31190 31496 D FileProcessor: Checking for tag: CameraOwnerName
06-25 12:38:04.853 31190 31495 D FileProcessor: Checking for tag: Copyright
06-25 12:38:04.854 31190 31507 D FileProcessor: Checking for tag: BodySerialNumber
06-25 12:38:04.854 31190 31502 D FileMetadataUtils: Retuning tags for pref:Software
06-25 12:38:04.857 31190 31508 D FileProcessor: Checking for tag: Make
06-25 12:38:04.858 31190 31495 D FileMetadataUtils: Retuning tags for pref:Camera Model
06-25 12:38:04.858 31190 31507 D FileProcessor: Checking for tag: LensSerialNumber
06-25 12:38:04.859 31190 31497 D FileMetadataUtils: Retuning tags for pref:Author
06-25 12:38:04.859 31190 31502 D FileProcessor: Checking for tag: Software
06-25 12:38:04.859 31190 31505 D CookieManager: Domain:commons.wikimedia.org
06-25 12:38:04.866 31190 31507 D FileMetadataUtils: Retuning tags for pref:Software
06-25 12:38:04.866 31190 31496 D FileMetadataUtils: Retuning tags for pref:Copyright
06-25 12:38:04.866 31190 31507 D FileProcessor: Checking for tag: Software
06-25 12:38:04.867 31190 31496 D FileProcessor: Checking for tag: Copyright
06-25 12:38:04.867 31190 31495 D FileProcessor: Checking for tag: Make
06-25 12:38:04.869 31190 31496 D FileMetadataUtils: Retuning tags for pref:Camera Model
06-25 12:38:04.869 31190 31508 D FileProcessor: Checking for tag: Model
06-25 12:38:04.872 31190 31497 D FileProcessor: Checking for tag: Artist
06-25 12:38:04.873 31190 31503 I FileUtils: File SHA1: 1dcb033455e6388a5767431d845d0730365373d8
06-25 12:38:04.873 31190 31497 D FileProcessor: Checking for tag: CameraOwnerName
06-25 12:38:04.873 31190 31495 D FileProcessor: Checking for tag: Model
06-25 12:38:04.873 31190 31508 D FileMetadataUtils: Retuning tags for pref:Lens Model
06-25 12:38:04.876 31190 31503 D OkHttp  : --> GET https://commons.wikimedia.org/w/api.php?action=query&format=json&formatversion=2&list=allimages&aisha1=1dcb033455e6388a5767431d845d0730365373d8
06-25 12:38:04.876 31190 31505 D OkHttp  : <-- HTTP FAILED: java.net.UnknownHostException: Unable to resolve host "commons.wikimedia.org": No address associated with hostname
06-25 12:38:04.876 31190 31496 D FileProcessor: Checking for tag: Make
06-25 12:38:04.877 31190 31508 D FileProcessor: Checking for tag: LensMake
06-25 12:38:04.877 31190 31495 D FileMetadataUtils: Retuning tags for pref:Lens Model
06-25 12:38:04.879 31190 31503 D CookieManager: Domain:commons.wikimedia.org
06-25 12:38:04.879 31190 31495 D FileProcessor: Checking for tag: LensMake
06-25 12:38:04.879 31190 31496 D FileProcessor: Checking for tag: Model
06-25 12:38:04.879 31190 31503 D OkHttp  : <-- HTTP FAILED: java.net.UnknownHostException: Unable to resolve host "commons.wikimedia.org": No address associated with hostname
06-25 12:38:04.879 31190 31497 D FileMetadataUtils: Retuning tags for pref:Copyright
06-25 12:38:04.880 31190 31495 D FileProcessor: Checking for tag: LensModel
06-25 12:38:04.881 31190 31497 D FileProcessor: Checking for tag: Copyright
06-25 12:38:04.882 31190 31496 D FileMetadataUtils: Retuning tags for pref:Lens Model
06-25 12:38:04.883 31190 31508 D FileProcessor: Checking for tag: LensModel
06-25 12:38:04.883 31190 31495 D FileProcessor: Checking for tag: LensSpecification
06-25 12:38:04.886 31190 31496 D FileProcessor: Checking for tag: LensMake
06-25 12:38:04.886 31190 31496 D FileProcessor: Checking for tag: LensModel
06-25 12:38:04.886 31190 31495 D FileMetadataUtils: Retuning tags for pref:Serial Numbers
06-25 12:38:04.887 31190 31497 D FileMetadataUtils: Retuning tags for pref:Camera Model
06-25 12:38:04.887 31190 31496 D FileProcessor: Checking for tag: LensSpecification
06-25 12:38:04.887 31190 31496 D FileMetadataUtils: Retuning tags for pref:Serial Numbers
06-25 12:38:04.890 31190 31497 D FileProcessor: Checking for tag: Make
06-25 12:38:04.890 31190 31507 I FileUtils: File SHA1: 877d7f7d509166ab968d02832bd6f4015f505071
06-25 12:38:04.890 31190 31497 D FileProcessor: Checking for tag: Model
06-25 12:38:04.891 31190 31508 D FileProcessor: Checking for tag: LensSpecification
06-25 12:38:04.891 31190 31497 D FileMetadataUtils: Retuning tags for pref:Lens Model
06-25 12:38:04.891 31190 31497 D FileProcessor: Checking for tag: LensMake
06-25 12:38:04.894 31190 31497 D FileProcessor: Checking for tag: LensModel
06-25 12:38:04.894 31190 31507 D OkHttp  : --> GET https://commons.wikimedia.org/w/api.php?action=query&format=json&formatversion=2&list=allimages&aisha1=877d7f7d509166ab968d02832bd6f4015f505071
06-25 12:38:04.895 31190 31495 D FileProcessor: Checking for tag: BodySerialNumber
06-25 12:38:04.895 31190 31508 D FileMetadataUtils: Retuning tags for pref:Serial Numbers
06-25 12:38:04.898 31190 31495 D FileProcessor: Checking for tag: LensSerialNumber
06-25 12:38:04.898 31190 31496 D FileProcessor: Checking for tag: BodySerialNumber
06-25 12:38:04.898 31190 31495 D FileMetadataUtils: Retuning tags for pref:Software
06-25 12:38:04.898 31190 31508 D FileProcessor: Checking for tag: BodySerialNumber
06-25 12:38:04.899 31190 31495 D FileProcessor: Checking for tag: Software
06-25 12:38:04.899 31190 31496 D FileProcessor: Checking for tag: LensSerialNumber
06-25 12:38:04.899 31190 31508 D FileProcessor: Checking for tag: LensSerialNumber
06-25 12:38:04.900 31190 31497 D FileProcessor: Checking for tag: LensSpecification
06-25 12:38:04.902 31190 31507 D CookieManager: Domain:commons.wikimedia.org
06-25 12:38:04.903 31190 31505 I FileUtils: File SHA1: 2865c1b3f4eefd18a2dd0371a833e4f729a4796b
06-25 12:38:04.904 31190 31496 D FileMetadataUtils: Retuning tags for pref:Software
06-25 12:38:04.904 31190 31503 I FileUtils: File SHA1: 8003a760c2ee8283165ac6f8ac9402d210551aa6
06-25 12:38:04.908 31190 31502 I FileUtils: File SHA1: 199c7009ee437ba6af0712942a0aeff6509aa197
06-25 12:38:04.908 31190 31507 D OkHttp  : <-- HTTP FAILED: java.net.UnknownHostException: Unable to resolve host "commons.wikimedia.org": No address associated with hostname
06-25 12:38:04.908 31190 31497 D FileMetadataUtils: Retuning tags for pref:Serial Numbers
06-25 12:38:04.909 31190 31508 D FileMetadataUtils: Retuning tags for pref:Software
06-25 12:38:04.910 31190 31496 D FileProcessor: Checking for tag: Software
06-25 12:38:04.911 31190 31508 D FileProcessor: Checking for tag: Software
06-25 12:38:04.912 31190 31497 D FileProcessor: Checking for tag: BodySerialNumber
06-25 12:38:04.914 31190 31502 D OkHttp  : --> GET https://commons.wikimedia.org/w/api.php?action=query&format=json&formatversion=2&list=allimages&aisha1=199c7009ee437ba6af0712942a0aeff6509aa197
06-25 12:38:04.918 31190 31497 D FileProcessor: Checking for tag: LensSerialNumber
06-25 12:38:04.921 31190 31502 D CookieManager: Domain:commons.wikimedia.org
06-25 12:38:04.922 31190 31497 D FileMetadataUtils: Retuning tags for pref:Software
06-25 12:38:04.934 31190 31502 D OkHttp  : <-- HTTP FAILED: java.net.UnknownHostException: Unable to resolve host "commons.wikimedia.org": No address associated with hostname
06-25 12:38:04.938 31190 31497 D FileProcessor: Checking for tag: Software
06-25 12:38:04.946 31190 31508 I FileUtils: File SHA1: 1388a7eae6ff489388787cfcc24f42babc1bbd21
06-25 12:38:04.947 31190 31508 D OkHttp  : --> GET https://commons.wikimedia.org/w/api.php?action=query&format=json&formatversion=2&list=allimages&aisha1=1388a7eae6ff489388787cfcc24f42babc1bbd21
06-25 12:38:04.948 31190 31507 I FileUtils: File SHA1: 6b48f480f50e8ee4f51485582961a4dff34e9eb5
06-25 12:38:04.949 31190 31505 D FileMetadataUtils: Retuning tags for pref:Author
06-25 12:38:04.950 31190 31496 I FileUtils: File SHA1: d30e28fe6d1422a3988fab9689c1dbf96aaaca17
06-25 12:38:04.950 31190 31508 D CookieManager: Domain:commons.wikimedia.org
06-25 12:38:04.950 31190 31505 D FileProcessor: Checking for tag: Artist
06-25 12:38:04.950 31190 31495 I FileUtils: File SHA1: 0197c876b8f6de3932ca0c455b6e5da73167a10e
06-25 12:38:04.950 31190 31505 D FileProcessor: Checking for tag: CameraOwnerName
06-25 12:38:04.950 31190 31508 D OkHttp  : <-- HTTP FAILED: java.net.UnknownHostException: Unable to resolve host "commons.wikimedia.org": No address associated with hostname
06-25 12:38:04.951 31190 31495 D OkHttp  : --> GET https://commons.wikimedia.org/w/api.php?action=query&format=json&formatversion=2&list=allimages&aisha1=0197c876b8f6de3932ca0c455b6e5da73167a10e
06-25 12:38:04.952 31190 31496 D OkHttp  : --> GET https://commons.wikimedia.org/w/api.php?action=query&format=json&formatversion=2&list=allimages&aisha1=d30e28fe6d1422a3988fab9689c1dbf96aaaca17
06-25 12:38:04.952 31190 31505 D FileMetadataUtils: Retuning tags for pref:Copyright
06-25 12:38:04.953 31190 31503 D FileMetadataUtils: Retuning tags for pref:Author
06-25 12:38:04.954 31190 31505 D FileProcessor: Checking for tag: Copyright
06-25 12:38:04.955 31190 31503 D FileProcessor: Checking for tag: Artist
06-25 12:38:04.958 31190 31495 D CookieManager: Domain:commons.wikimedia.org
06-25 12:38:04.958 31190 31505 D FileMetadataUtils: Retuning tags for pref:Camera Model
06-25 12:38:04.962 31190 31496 D CookieManager: Domain:commons.wikimedia.org
06-25 12:38:04.964 31190 31503 D FileProcessor: Checking for tag: CameraOwnerName
06-25 12:38:04.965 31190 31505 D FileProcessor: Checking for tag: Make
06-25 12:38:04.965 31190 31503 D FileMetadataUtils: Retuning tags for pref:Copyright
06-25 12:38:04.965 31190 31495 D OkHttp  : <-- HTTP FAILED: java.net.UnknownHostException: Unable to resolve host "commons.wikimedia.org": No address associated with hostname
06-25 12:38:04.965 31190 31503 D FileProcessor: Checking for tag: Copyright
06-25 12:38:04.966 31190 31496 D OkHttp  : <-- HTTP FAILED: java.net.UnknownHostException: Unable to resolve host "commons.wikimedia.org": No address associated with hostname
06-25 12:38:04.970 31190 31505 D FileProcessor: Checking for tag: Model
06-25 12:38:04.970 31190 31503 D FileMetadataUtils: Retuning tags for pref:Camera Model
06-25 12:38:04.970 31190 31503 D FileProcessor: Checking for tag: Make
06-25 12:38:04.971 31190 31503 D FileProcessor: Checking for tag: Model
06-25 12:38:04.971 31190 31505 D FileMetadataUtils: Retuning tags for pref:Lens Model
06-25 12:38:04.971 31190 31503 D FileMetadataUtils: Retuning tags for pref:Lens Model
06-25 12:38:04.972 31190 31503 D FileProcessor: Checking for tag: LensMake
06-25 12:38:04.972 31190 31508 I FileUtils: File SHA1: d5c69e57199d5716b78f20cef06a3eb81e66fc2c
06-25 12:38:04.972 31190 31505 D FileProcessor: Checking for tag: LensMake
06-25 12:38:04.972 31190 31503 D FileProcessor: Checking for tag: LensModel
06-25 12:38:04.972 31190 31503 D FileProcessor: Checking for tag: LensSpecification
06-25 12:38:04.973 31190 31503 D FileMetadataUtils: Retuning tags for pref:Serial Numbers
06-25 12:38:04.973 31190 31505 D FileProcessor: Checking for tag: LensModel
06-25 12:38:04.974 31190 31503 D FileProcessor: Checking for tag: BodySerialNumber
06-25 12:38:04.974 31190 31503 D FileProcessor: Checking for tag: LensSerialNumber
06-25 12:38:04.974 31190 31503 D FileMetadataUtils: Retuning tags for pref:Software
06-25 12:38:04.975 31190 31503 D FileProcessor: Checking for tag: Software
06-25 12:38:04.977 31190 31502 I FileUtils: File SHA1: cdeb61fa1f94c2da68475644ececdc59ad38c14e
06-25 12:38:04.979 31190 31497 I FileUtils: File SHA1: b46edb33ed6fdd73a7e4e705032bca19f697af75
06-25 12:38:04.981 31190 31503 I FileUtils: File SHA1: 19c2ec6ce62ce7c92c048be79f932ccdce9904d4
06-25 12:38:04.982 31190 31505 D FileProcessor: Checking for tag: LensSpecification
06-25 12:38:04.982 31190 31497 D OkHttp  : --> GET https://commons.wikimedia.org/w/api.php?action=query&format=json&formatversion=2&list=allimages&aisha1=b46edb33ed6fdd73a7e4e705032bca19f697af75
06-25 12:38:04.990 31190 31503 D OkHttp  : --> GET https://commons.wikimedia.org/w/api.php?action=query&format=json&formatversion=2&list=allimages&aisha1=19c2ec6ce62ce7c92c048be79f932ccdce9904d4
06-25 12:38:04.990 31190 31497 D CookieManager: Domain:commons.wikimedia.org
06-25 12:38:04.992 31190 31505 D FileMetadataUtils: Retuning tags for pref:Serial Numbers
06-25 12:38:04.992 31190 31497 D OkHttp  : <-- HTTP FAILED: java.net.UnknownHostException: Unable to resolve host "commons.wikimedia.org": No address associated with hostname
06-25 12:38:04.994 31190 31503 D CookieManager: Domain:commons.wikimedia.org
06-25 12:38:04.994 31190 31503 D OkHttp  : <-- HTTP FAILED: java.net.UnknownHostException: Unable to resolve host "commons.wikimedia.org": No address associated with hostname
06-25 12:38:04.995 31190 31505 D FileProcessor: Checking for tag: BodySerialNumber
06-25 12:38:04.998 31190 31505 D FileProcessor: Checking for tag: LensSerialNumber
06-25 12:38:04.998 31190 31505 D FileMetadataUtils: Retuning tags for pref:Software
06-25 12:38:04.998 31190 31505 D FileProcessor: Checking for tag: Software
06-25 12:38:05.000 31190 31507 D FileMetadataUtils: Retuning tags for pref:Author
06-25 12:38:05.001 31190 31507 D FileProcessor: Checking for tag: Artist
06-25 12:38:05.003 31190 31507 D FileProcessor: Checking for tag: CameraOwnerName
06-25 12:38:05.004 31190 31507 D FileMetadataUtils: Retuning tags for pref:Copyright
06-25 12:38:05.005 31190 31507 D FileProcessor: Checking for tag: Copyright
06-25 12:38:05.006 31190 31507 D FileMetadataUtils: Retuning tags for pref:Camera Model
06-25 12:38:05.006 31190 31507 D FileProcessor: Checking for tag: Make
06-25 12:38:05.007 31190 31507 D FileProcessor: Checking for tag: Model
06-25 12:38:05.007 31190 31507 D FileMetadataUtils: Retuning tags for pref:Lens Model
06-25 12:38:05.007 31190 31507 D FileProcessor: Checking for tag: LensMake
06-25 12:38:05.008 31190 31507 D FileProcessor: Checking for tag: LensModel
06-25 12:38:05.008 31190 31507 D FileProcessor: Checking for tag: LensSpecification
06-25 12:38:05.009 31190 31507 D FileMetadataUtils: Retuning tags for pref:Serial Numbers
06-25 12:38:05.009 31190 31507 D FileProcessor: Checking for tag: BodySerialNumber
06-25 12:38:05.010 31190 31507 D FileProcessor: Checking for tag: LensSerialNumber
06-25 12:38:05.010 31190 31507 D FileMetadataUtils: Retuning tags for pref:Software
06-25 12:38:05.010 31190 31507 D FileProcessor: Checking for tag: Software
06-25 12:38:05.011 31190 31505 I FileUtils: File SHA1: 97bec2355533ec96de94af0b7ac4684da20b2459
06-25 12:38:05.011 31190 31505 D OkHttp  : --> GET https://commons.wikimedia.org/w/api.php?action=query&format=json&formatversion=2&list=allimages&aisha1=97bec2355533ec96de94af0b7ac4684da20b2459
06-25 12:38:05.013 31190 31505 D CookieManager: Domain:commons.wikimedia.org
06-25 12:38:05.017 31190 31505 D OkHttp  : <-- HTTP FAILED: java.net.UnknownHostException: Unable to resolve host "commons.wikimedia.org": No address associated with hostname
06-25 12:38:05.018 31190 31201 I ree.nrw.common: Background concurrent copying GC freed 70856(5MB) AllocSpace objects, 8(544KB) LOS objects, 50% free, 17MB/34MB, paused 542us total 304.080ms
06-25 12:38:05.029 31190 31190 I Process : Sending signal. PID: 31190 SIG: 9
06-25 12:38:05.108  5124  5124 I Zygote  : Process 31190 exited due to signal (9)
06-25 12:38:05.109  5621  5664 I libprocessgroup: Successfully killed process cgroup uid 10401 pid 31190 in 0ms
```

4D17Y4 commented:
@nicolas-raoul 
That was due to an unhandled connection interrupt, handled it.
Needs further attention if we want to notify the user about it.

madhurgupta10 commented on app/src/main/java/fr/free/nrw/commons/customselector/ui/selector/ImageLoader.kt:
```diff
@@ -62,26 +74,30 @@ class ImageLoader @Inject constructor(
+                    val uploadedStatus = uploadedStatusDao.getUploadedFromImageSHA1(imageSHA1)
+                    var sha1 = ""
+
+                    if (uploadedStatus != null) {
```
use `let` here (same for line 111)

4D17Y4 replied:
Done.

madhurgupta10 commented on app/src/main/java/fr/free/nrw/commons/customselector/ui/selector/ImageLoader.kt:
```diff
@@ -91,13 +107,23 @@ class ImageLoader @Inject constructor(
      * @return Query result.
      */
-    private fun querySHA1(SHA1: String): Boolean {
+    private fun querySHA1(SHA1: String): Boolean? {
```
do we need to set this as nullable (same for line 78 & 114)?

4D17Y4 replied:
I'm using null to notify that the result was not available or it is invalid.

Should I declare some constants for this result instead of using a boolean as we want 3 states?

madhurgupta10 replied:
How about making a Sealed Class instead?

4D17Y4 replied:
Done. Thanks.

nicolas-raoul commented:
> Needs further attention if we want to notify the user about it.

No need to notify the user.

madhurgupta10 reviewed (approved):
Haven't tested the PR but changes look go to me :)

nicolas-raoul commented:
It works great!
I noticed a minor thing that I think could be improved:

Steps:
1. Reset app data
2. Start app
3. Log in
4. Set airplane mode
5. Use picker. The "uploaded" overlay is not shown on any picture, which is correct.
6. Leave airplane mode
7. Use picker. The "uploaded" overlay is not shown on any picture, which is problematic.

When an API request fails for any reason, it is good that the picker gives up and shows the picture as not uploaded, but the app should not record that in database. By doing that, the app can try again next time the picker is opened.

4D17Y4 commented:
Thanks, @nicolas-raoul, 
Added another object in the sealed item class, confirmed error before database insert.

nicolas-raoul commented:
Wonderful, it works great now!

4D17Y4 commented:
PR completed, up for testing and review.

New changes:
Updated the uploaded status table as soon as an upload is complete.

Note: Updated the contribution model but did not update the database version as we have already done that in this PR. Due to the change, the app might crash on start. A fresh install would fix it.

nicolas-raoul commented:
It works great, thanks!
